### ISSUE TITLE
Cannot connect to Amazon Redshift

### ISSUE THREAD

**Issue by peter** [ericmj/postgrex#107]
I don't know which versions of PostgreSQL the postgrex library is intended to support but I cannot get it to work with Amazon Redshift which is a modified version of PostgreSQL 8.0.2

http://docs.aws.amazon.com/redshift/latest/dg/c_redshift-and-postgres-sql.html

I get a number of different errors trying to connect to and query Redshift. See my console session below. I tried patching a few of those errors, such as patching Connection.query to deal with a response of :ok:

  def query(pid, statement, params, opts \\ []) do
    message = {:query, statement, params}
    timeout = opts[:timeout] || @timeout
    case GenServer.call(pid, message, timeout) do
      :ok ->
        {:ok, []}
      %Postgrex.Result{} = res ->
        {:ok, res}
      %Postgrex.Error{} = err ->
        {:error, err}
      {:error, kind, reason, stack} ->
        :erlang.raise(kind, reason, stack)
    end
  end

However that just triggered other problems down the line related to types and I don't know enough about Postgrex to fix them and I don't even know if Redshift should be supported. Should it? Is there any other way to connect to Redshift from Erlang/Elixir?

iex(1)> {:ok, pid} = Postgrex.Connection.start_link(hostname: "innovation-dw.cjer9kp9xlsu.eu-west-1.redshift.amazonaws.com", username: "<username-here>", password: "<password-here>", database: "<database-here>", port: 5439)
{:ok, #PID<0.449.0>}
iex(2)> [warn] Unhandled Postgres error: ERROR (syntax_error): syntax error at or near "]"

nil

iex(3)> Postgrex.Connection.query!(pid, "select * from foo", [])
** (EXIT from #PID<0.447.0>) an exception was raised:
    ** (MatchError) no match of right hand side value: ["bla"]
        (postgrex) lib/postgrex/types.ex:84: anonymous fn/1 in Postgrex.Types.build_types/1
        (elixir) lib/enum.ex:1043: anonymous fn/3 in Enum.map/2
        (elixir) lib/enum.ex:1387: Enum."-reduce/3-lists^foldl/2-0-"/3
        (elixir) lib/enum.ex:1043: Enum.map/2
        (postgrex) lib/postgrex/protocol.ex:159: Postgrex.Protocol.message/3
        (postgrex) lib/postgrex/connection.ex:417: Postgrex.Connection.new_data/2
        (postgrex) lib/postgrex/connection.ex:292: Postgrex.Connection.handle_info/2
        (stdlib) gen_server.erl:615: :gen_server.try_dispatch/4
        (stdlib) gen_server.erl:681: :gen_server.handle_msg/5
        (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3

Interactive Elixir (1.1.1) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> [error] GenServer #PID<0.449.0> terminating
** (MatchError) no match of right hand side value: ["bla"]
    (postgrex) lib/postgrex/types.ex:84: anonymous fn/1 in Postgrex.Types.build_types/1
    (elixir) lib/enum.ex:1043: anonymous fn/3 in Enum.map/2
    (elixir) lib/enum.ex:1387: Enum."-reduce/3-lists^foldl/2-0-"/3
    (elixir) lib/enum.ex:1043: Enum.map/2
    (postgrex) lib/postgrex/protocol.ex:159: Postgrex.Protocol.message/3
    (postgrex) lib/postgrex/connection.ex:417: Postgrex.Connection.new_data/2
    (postgrex) lib/postgrex/connection.ex:292: Postgrex.Connection.handle_info/2
    (stdlib) gen_server.erl:615: :gen_server.try_dispatch/4
    (stdlib) gen_server.erl:681: :gen_server.handle_msg/5
    (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Last message: {:tcp, #Port<0.13922>, <<50, 0, 0, 0, 4, 68, 0, 0, 0, 13, 0, 1, 0, 0, 0, 3, 98, 108, 97, 67, 0, 0, 0, 11, 83, 69, 76, 69, 67, 84, 0, 90, 0, 0, 0, 5, 73>>}
State: %{backend_key: {1354, 405653114}, bootstrap: {#Reference<0.0.2.1879>, 1384548, [Postgrex.Extensions.Binary, Postgrex.Extensions.Text], #HashDict<[{Postgrex.Extensions.Text, {8, 0, 2}}, {Postgrex.Extensions.Binary, {8, 0, 2}}]>}, extensions: [{Postgrex.Extensions.Binary, nil}, {Postgrex.Extensions.Text, nil}], listener_channels: #HashDict<[]>, listeners: #HashDict<[]>, opts: [password: :REDACTED, hostname: "innovation-dw.cjer9kp9xlsu.eu-west-1.redshift.amazonaws.com", username: "inno_dw", database: "dev", port: 5439], parameters: %{"TimeZone" => "UTC", "client_encoding" => "UNICODE", "datestyle" => "ISO, MDY", "gconf_case_sensitive" => "on", "integer_datetimes" => "on", "is_superuser" => "on", "max_numeric_precision" => "38", "max_varchar_size" => "65535", "padb_revision" => "V1-Patch61-44-g04ecee3f46", "padb_version" => "", "server_encoding" => "UNICODE", "server_version" => "8.0.2", "session_authorization" => "inno_dw", "timezone_abbreviations" => "Default"}, portal: [], queue: {[%{command: {:query, "select * from foo", []}, from: {#PID<0.447.0>, #Reference<0.0.1.3732>}, reply: :no_reply}], []}, rows: [], sock: {:gen_tcp, #Port<0.13922>}, state: :binding, statement: nil, tail: "", types: :types_removed, types_key: {'innovation-dw.cjer9kp9xlsu.eu-west-1.redshift.amazonaws.com', 5439, "dev", []}}

iex(10)> Postgrex.Connection.query!(pid, "select true", [])
** (EXIT from #PID<0.454.0>) an exception was raised:
    ** (MatchError) no match of right hand side value: ["t"]
        (postgrex) lib/postgrex/types.ex:84: anonymous fn/1 in Postgrex.Types.build_types/1
        (elixir) lib/enum.ex:1043: anonymous fn/3 in Enum.map/2
        (elixir) lib/enum.ex:1387: Enum."-reduce/3-lists^foldl/2-0-"/3
        (elixir) lib/enum.ex:1043: Enum.map/2
        (postgrex) lib/postgrex/protocol.ex:159: Postgrex.Protocol.message/3
        (postgrex) lib/postgrex/connection.ex:417: Postgrex.Connection.new_data/2
        (postgrex) lib/postgrex/connection.ex:292: Postgrex.Connection.handle_info/2
        (stdlib) gen_server.erl:615: :gen_server.try_dispatch/4
        (stdlib) gen_server.erl:681: :gen_server.handle_msg/5
        (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3

Interactive Elixir (1.1.1) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> [error] GenServer #PID<0.461.0> terminating
** (MatchError) no match of right hand side value: ["t"]
    (postgrex) lib/postgrex/types.ex:84: anonymous fn/1 in Postgrex.Types.build_types/1
    (elixir) lib/enum.ex:1043: anonymous fn/3 in Enum.map/2
    (elixir) lib/enum.ex:1387: Enum."-reduce/3-lists^foldl/2-0-"/3
    (elixir) lib/enum.ex:1043: Enum.map/2
    (postgrex) lib/postgrex/protocol.ex:159: Postgrex.Protocol.message/3
    (postgrex) lib/postgrex/connection.ex:417: Postgrex.Connection.new_data/2
    (postgrex) lib/postgrex/connection.ex:292: Postgrex.Connection.handle_info/2
    (stdlib) gen_server.erl:615: :gen_server.try_dispatch/4
    (stdlib) gen_server.erl:681: :gen_server.handle_msg/5
    (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Last message: {:tcp, #Port<0.14244>, <<50, 0, 0, 0, 4, 68, 0, 0, 0, 11, 0, 1, 0, 0, 0, 1, 116, 67, 0, 0, 0, 11, 83, 69, 76, 69, 67, 84, 0, 90, 0, 0, 0, 5, 73>>}
State: %{backend_key: {1439, 1808437079}, bootstrap: {#Reference<0.0.1.3793>, 1388644, [Postgrex.Extensions.Binary, Postgrex.Extensions.Text], #HashDict<[{Postgrex.Extensions.Text, {8, 0, 2}}, {Postgrex.Extensions.Binary, {8, 0, 2}}]>}, extensions: [{Postgrex.Extensions.Binary, nil}, {Postgrex.Extensions.Text, nil}], listener_channels: #HashDict<[]>, listeners: #HashDict<[]>, opts: [password: :REDACTED, hostname: "innovation-dw.cjer9kp9xlsu.eu-west-1.redshift.amazonaws.com", username: "inno_dw", database: "dev", port: 5439], parameters: %{"TimeZone" => "UTC", "client_encoding" => "UNICODE", "datestyle" => "ISO, MDY", "gconf_case_sensitive" => "on", "integer_datetimes" => "on", "is_superuser" => "on", "max_numeric_precision" => "38", "max_varchar_size" => "65535", "padb_revision" => "V1-Patch61-44-g04ecee3f46", "padb_version" => "", "server_encoding" => "UNICODE", "server_version" => "8.0.2", "session_authorization" => "inno_dw", "timezone_abbreviations" => "Default"}, portal: [], queue: {[%{command: {:query, "select true", []}, from: {#PID<0.454.0>, #Reference<0.0.1.3810>}, reply: :no_reply}], []}, rows: [], sock: {:gen_tcp, #Port<0.14244>}, state: :binding, statement: nil, tail: "", types: :types_removed, types_key: {'innovation-dw.cjer9kp9xlsu.eu-west-1.redshift.amazonaws.com', 5439, "dev", []}}

nil

**Comment by michalmuskala** [comment]
According to readme 8.0 is not supported, so this may be the reason for errors:
> Supports PostgreSQL 8.4, 9.0, 9.1, 9.2, 9.3, and 9.4 (hstore is not supported on 8.4)

**Comment by peter** [comment]
Cool, thanks, I missed the section on supported versions in the README, sorry.

**Comment by colinbankier** [comment]
Hi Peter,
I'm currently using postgrex with Redshift in a production phoenix app. I forked postgrex to make it work here: https://github.com/colinbankier/postgrex (master branch)
Disclaimer: I implemented what I needed to get things working for myself, and didn't intend have it used elsewhere without some further work :)
The main issue was with bootstrapping types, since Redshift supports only a subset of postgres types, and doesn't support some syntax for arrays etc.
Our app uses both a redshift and a postgres Repo, so I needed to support also talking to postgres with the bootstrap process "unadulterated".
To use it in "Redshift compatible" mode, just add 

      bootstrap_module: Postgrex.Bootstrap.Redshift

to your repo config.
I would love some help getting this in a form that could be brought back into the main project :)
